homeassistant/components/generic/config_flow.py

```diff
@@ -0,0 +1,318 @@
+from homeassistant.helpers import config_validation as cv, template as template_helper
+from homeassistant.helpers.httpx_client import get_async_client
+
+from .camera import generate_auth
+
+# pylint: disable=unused-import
+from .const import (
+    CONF_CONTENT_TYPE,
+    CONF_FRAMERATE,
+    CONF_LIMIT_REFETCH_TO_URL_CHANGE,
+    CONF_RTSP_TRANSPORT,
+    CONF_STILL_IMAGE_URL,
+    CONF_STREAM_SOURCE,
+    DEFAULT_NAME,
+    DOMAIN,
+    FFMPEG_OPTION_MAP,
+    GET_IMAGE_TIMEOUT,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+DEFAULT_DATA = {
+    CONF_NAME: DEFAULT_NAME,
+    CONF_AUTHENTICATION: HTTP_BASIC_AUTHENTICATION,
+    CONF_LIMIT_REFETCH_TO_URL_CHANGE: False,
+    CONF_FRAMERATE: 2,
+    CONF_VERIFY_SSL: True,
+}
+
+SUPPORTED_IMAGE_TYPES = ["png", "jpeg", "svg+xml"]
+
+
+def build_schema(user_input):
+    """Create schema for camera config setup."""
+    spec = {
+        vol.Optional(
+            CONF_STILL_IMAGE_URL,
+            description={"suggested_value": user_input.get(CONF_STILL_IMAGE_URL, "")},
+        ): str,
+        vol.Optional(
+            CONF_STREAM_SOURCE,
+            description={"suggested_value": user_input.get(CONF_STREAM_SOURCE, "")},
+        ): str,
+        vol.Optional(
+            CONF_RTSP_TRANSPORT, default=user_input.get(CONF_RTSP_TRANSPORT)
+        ): vol.In([None, "tcp", "udp", "udp_multicast", "http"]),
+        vol.Optional(
+            CONF_AUTHENTICATION, default=user_input.get(CONF_AUTHENTICATION)
+        ): vol.In([HTTP_BASIC_AUTHENTICATION, HTTP_DIGEST_AUTHENTICATION]),
+        vol.Optional(
+            CONF_USERNAME,
+            description={"suggested_value": user_input.get(CONF_USERNAME, "")},
+        ): str,
+        vol.Optional(
+            CONF_PASSWORD,
+            description={"suggested_value": user_input.get(CONF_PASSWORD, "")},
+        ): str,
+        vol.Required(
+            CONF_LIMIT_REFETCH_TO_URL_CHANGE,
+            default=user_input.get(CONF_LIMIT_REFETCH_TO_URL_CHANGE, False),
+        ): bool,
+        vol.Required(
+            CONF_FRAMERATE,
+            description={"suggested_value": user_input.get(CONF_FRAMERATE, 2)},
+        ): int,
+        vol.Required(
+            CONF_VERIFY_SSL, default=user_input.get(CONF_VERIFY_SSL, True)
+        ): bool,
+    }
+    return vol.Schema(spec)
+
+
+def get_image_type(image):
+    """Get the format of downloaded bytes that could be an image."""
+    fmt = imghdr.what(None, h=image)
+    if fmt is None:
+        # if imghdr can't figure it out, could be svg.
+        try:
+            if image.decode("utf-8").startswith("<svg"):
+                return "svg+xml"
+        except UnicodeDecodeError:
+            return None
+    return fmt
+
+
+async def async_test_connection(
+    hass, info
+) -> tuple[dict[str, str], str | None, str | None]:
+    """Verify that the connection data is valid before we add it."""
+    errors: dict[str, str] = {}
+    fmt = None
+    if url := info.get(CONF_STILL_IMAGE_URL):
```

Comment:
This would be cleaner if there was a separate function to validate the image url, and one to validate the stream source.  Having everything in one function makes this quite complex.

Reply:
Ok, done. f269a8e19

Reply:
You can save indent here as well per suggestion https://github.com/home-assistant/core/pull/52360/files#r815247034